### PR TITLE
Consistency Overlord

### PR DESCRIPTION
**Pull request by freesig**
This PR introduces a protocol to check for consistency in a sharded dht and provides a implementation for multiple conductors run in the same test.
A follow up will be to implement the protocol in the rust bootstrap to allow consistency to be checked across multiple machines.
This protocol makes no assumptions on how the data is used and implementations could aim for a statistical level of consistency over strict consistency. It also aims to keep network traffic to a minimum by allowing the expected data to be gathered at the start of the session and then having agents self check there own status. Agents then report back their status at rates that can be configured.
For a network implementation I would suggest staggering the reporting rates to avoid network stampedes.

### CODE REVIEW

File: crates/holochain/src/test_utils/consistency.rs [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+                    "
+                    SELECT
+                    DhtOp.hash as dht_op_hash
+                    FROM DhtOp
+                    JOIN
+                    Header ON DhtOp.header_hash = Header.hash
+                    WHERE
+                    DhtOp.is_authored = 1
+                    AND
+                    (DhtOp.type != :store_entry OR Header.private_entry = 0)
```
Comment (maackle) [inline crates/holochain/src/test_utils/consistency.rs:457]: Just curious, is the DhtOp.type check redundant? Can something be private that's not a StoreEntry?

File: crates/holochain/src/test_utils/consistency.rs [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+            },
+    } = session;
+
+    let start = tokio::time::Instant::now();
+    let mut last_keep_alive = start;
+    let deadline = tokio::time::Instant::now() + timeout;
+    let mut frequency = tokio::time::interval(frequency);
+    let mut missing_agents = Vec::with_capacity(expected_agents.len());
+    let mut missing_hashes = Vec::with_capacity(expected_hashes.len());
+    while let Ok(_) = tokio::time::timeout_at(deadline, frequency.tick()).await {
```
Comment (maackle) [inline crates/holochain/src/test_utils/consistency.rs:332]: Wouldn't this `deadline` timeout be applied independently to every tick, not to the overall while loop?

File: crates/holochain/src/test_utils/consistency.rs [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+}
+
+async fn wait_for_consistency(
+    mut rx: tokio::sync::mpsc::Receiver<SessionMessage>,
+    mut agents: HashSet<Arc<KitsuneAgent>>,
+    timeout: Duration,
+) {
+    let deadline = tokio::time::Instant::now() + timeout + Duration::from_millis(100);
+    while let Ok(Some(SessionMessage { from, report })) =
+        tokio::time::timeout_at(deadline, rx.recv()).await
```
Comment (maackle) [inline crates/holochain/src/test_utils/consistency.rs:152]: Also not sure about this use of `deadline`, I would expect the deadline to be the deadline for achieving total consistency, i.e. across receiving all reports, not just each one.

File: crates/holochain/src/test_utils/consistency.rs (hunk 1 above)
Comment (freesig) [reply crates/holochain/src/test_utils/consistency.rs:457]: Yes but we publish private ops that aren't the StoreEntry

File: crates/holochain/src/test_utils/consistency.rs (hunk 2 above)
Comment (freesig) [reply crates/holochain/src/test_utils/consistency.rs:332]: A deadline is a point in time that when we've past any future will timeout immediately. It's not relative to the time the future was called.

File: crates/holochain/src/test_utils/consistency.rs (hunk 3 above)
Comment (freesig) [reply crates/holochain/src/test_utils/consistency.rs:152]: The deadline is created at the start of the loop so is the same instant for all `rx.recv()` and will timeout the loop if we haven't successfully exited.

**Review by steveeJ — APPROVED** [review]
@freesig walked me through these changes -- thanks for that! -- and despite a misnomer of some of the fields in `SessionReport::KeepAlive` it makes sense to me. 

we also discussed whether we could use something like Prometheus to collect the kinds of metrics that are being created here without having reached a conclusion on that just yet.

**Review by maackle — APPROVED** [review]

**Pull request merged**